File: Estruturas de Dados e Complexidade de Algoritmos/Algoritmos12/Alg12.java
/*
♦ Algoritmo 12: Estrutura de Dados Linear Dinâmica (não Estática) de Lista Ligada/Encadeada Simples (não Duplamente Ligada): Linked List (Lista Ligada).
*/

public class Alg12<T> {

    private Nodo<T> primeiro;
    private int tamanho;

    public Alg12() {
        primeiro = null;
        tamanho = 0;
    }

    public void addList(T valor) {
        Nodo<T> novoNodo = new Nodo<>(valor);
        novoNodo.setProximo(primeiro);
        primeiro = novoNodo;
        tamanho++;
    }

    public void removeList(T valor) {
        if (primeiro == null) {
            return;
        }

        if (primeiro.getValor().equals(valor)) {
            primeiro = primeiro.getProximo();
            tamanho--;
            return;
        }

        Nodo<T> nodoAtual = primeiro;
        while (nodoAtual.getProximo() != null) {
            if (nodoAtual.getProximo().getValor().equals(valor)) {
                nodoAtual.setProximo(nodoAtual.getProximo().getProximo());
                tamanho--;
                return;
            }
            nodoAtual = nodoAtual.getProximo();
        }
    }

    public int size() {
        return tamanho;
    }

    public void showList() {
        Nodo<T> nodoAtual = primeiro;
        while (nodoAtual != null) {
            System.out.print(nodoAtual.getValor() + " ");
            nodoAtual = nodoAtual.getProximo();
        }
        System.out.println();
    }

    public static void main(String[] args) {
        Alg12<Integer> lista = new Alg12<>();
        lista.addList(10);
        lista.addList(20);
        lista.addList(30);

        System.out.println("Tamanho da lista: " + lista.size());
        System.out.print("Elementos da lista: ");
        lista.showList();

        lista.removeList(20);

        System.out.println("Tamanho da lista após remoção: " + lista.size());
        System.out.print("Elementos da lista após remoção: ");
        lista.showList();
    }
}

class Nodo<T> {
    private T valor;
    private Nodo<T> proximo;

    public Nodo(T valor) {
        this.valor = valor;
        this.proximo = null;
    }

    public T getValor() {
        return valor;
    }

    public Nodo<T> getProximo() {
        return proximo;
    }

    public void setProximo(Nodo<T> proximo) {
        this.proximo = proximo;
    }
}

File: Estruturas de Dados e Complexidade de Algoritmos/Algoritmos12/Alg12.rb
=begin
♦ Algoritmo 12: Estrutura de Dados Linear Dinâmica (não Estática) de Lista Ligada/Encadeada Simples (não Duplamente Ligada): Linked List (Lista Ligada).
=end

class Nodo
  attr_accessor :valor, :proximo

  def initialize(valor)
    @valor = valor
    @proximo = nil
  end
end

class ListaLigada
  def initialize
    @primeiro = nil
    @tamanho = 0
  end

  def add_list(valor)
    novo_nodo = Nodo.new(valor)
    novo_nodo.proximo = @primeiro
    @primeiro = novo_nodo
    @tamanho += 1
  end

  def remove_list(valor)
    return if @primeiro.nil?

    if @primeiro.valor == valor
      @primeiro = @primeiro.proximo
      @tamanho -= 1
      return
    end

    nodo_atual = @primeiro
    while nodo_atual.proximo
      if nodo_atual.proximo.valor == valor
        nodo_atual.proximo = nodo_atual.proximo.proximo
        @tamanho -= 1
        return
      end
      nodo_atual = nodo_atual.proximo
    end
  end

  def size
    @tamanho
  end

  def show_list
    nodo_atual = @primeiro
    while nodo_atual
      print nodo_atual.valor.to_s + ' '
      nodo_atual = nodo_atual.proximo
    end
    puts
  end
end

lista = ListaLigada.new
lista.add_list(10)
lista.add_list(20)
lista.add_list(30)

puts 'Tamanho da lista:', lista.size
print 'Elementos da lista: '
lista.show_list

lista.remove_list(20)

puts 'Tamanho da lista após remoção:', lista.size
print 'Elementos da lista após remoção: '
lista.show_list
